marlowe-runtime/src/Language/Marlowe/Runtime/History/Follower.hs

```diff
@@ -0,0 +1,559 @@
+  let TransactionScriptOutput utxo prevDatum = scriptOutput
+  marloweTx@Transaction{output} <- lift
+    $ first ExtractMarloweTransactionFailed
+    $ extractMarloweTransaction version prevDatum slotConfig contractId scriptAddress payoutValidatorHash utxo blockHeader tx
+  tellStep blockHeader $ ApplyTransaction marloweTx
+  pure output
+
+tellStep :: BlockHeader -> ContractStep v -> WriterT (ContractChanges v) (Either ContractHistoryError) ()
+tellStep blockHeader step = tell mempty { steps = Map.singleton blockHeader [step] }
+
+extractMarloweTransaction
+  :: MarloweVersion v
+  -> Datum v
+  -> SlotConfig
+  -> ContractId
+  -> Chain.Address
+  -> Chain.ScriptHash
+  -> TxOutRef
+  -> BlockHeader
+  -> Chain.Transaction
+  -> Either ExtractMarloweTransactionError (Transaction v)
+extractMarloweTransaction version prevDatum slotConfig contractId scriptAddress payoutValidatorHash consumedUTxO blockHeader Chain.Transaction{..} = do
+  let transactionId = txId
+  Chain.TransactionInput { redeemer = mRedeemer } <-
+    note TxInNotFound $ find (consumesUTxO consumedUTxO) inputs
+  rawRedeemer <- note NoRedeemer mRedeemer
+  redeemer <- note InvalidRedeemer $ fromChainRedeemer version rawRedeemer
+  (minSlot, maxSlot) <- case validityRange of
+    Chain.MinMaxBound minSlot maxSlot -> pure (minSlot, maxSlot)
+    _                                 -> Left InvalidValidityRange
+  let validityLowerBound = slotToUTCTime slotConfig minSlot
+  let validityUpperBound = slotToUTCTime slotConfig maxSlot
+  let wouldClose = wouldCloseContract version validityLowerBound validityUpperBound redeemer prevDatum
+  scriptOutput <- runMaybeT do
+    guard $ not wouldClose
+    (ix, Chain.TransactionOutput{ datum = mDatum }) <-
+      hoistMaybe $ find (isToAddress scriptAddress . snd) $ zip [0..] outputs
+    lift do
+      rawDatum <- note NoTransactionDatum mDatum
+      datum <- note InvalidTransactionDatum $ fromChainDatum version rawDatum
+      let txIx = Chain.TxIx ix
+      let utxo = Chain.TxOutRef{..}
+      pure TransactionScriptOutput{..}
+  let
+    payoutOutputs = Map.filter (isToScriptHash payoutValidatorHash)
+      $ Map.fromList
+      $ (\(txIx, output) -> (Chain.TxOutRef{..}, output)) <$> zip [0..] outputs
+  payouts <- flip Map.traverseWithKey payoutOutputs \txOut Chain.TransactionOutput{datum=mPayoutDatum, assets} -> do
+    rawPayoutDatum <- note (NoPayoutDatum txOut) mPayoutDatum
+    payoutDatum <- note (InvalidPayoutDatum txOut) $ fromChainPayoutDatum version rawPayoutDatum
+    pure $ Payout assets payoutDatum
+  let output = TransactionOutput{..}
+  pure Transaction{..}
+
+wouldCloseContract :: MarloweVersion v -> UTCTime -> UTCTime -> Redeemer v -> Datum v -> Bool
+wouldCloseContract version validityLowerBound validityUpperBound redeemer datum = case version of
+  MarloweV1 ->
+    let
+      utcTimeToPOSIXTime = POSIXTime . round . (* 1000) . utcTimeToPOSIXSeconds
```

Comment:
I think this is okay, but we have to be careful about rounding milliseconds in times. (We've previously encountered sub-millisecond failures in `computeTransaction` and in the validator because of this.) Let's make sure we eventually have a thorough set of tests for times and slots. Also, we might want to consider debug logging failures from `computeTransaction` so these time-skew failures don't slip by us.

Reply:
Yeah, they should be logged as warnings